joomla/joomla-cms
# [6.1] Workflow emails

Pull Request for Issue #37745 #37446 .

### Summary of Changes
This change adds the option to select email as a notification type. Users can send notifications to:
- Private messages (old and default behavior)
- Email

![image](https://user-images.githubusercontent.com/359377/169713511-d2a97857-3a5c-4e29-8899-c2fd38b111b4.png)


### Testing Instructions
First you need to apply this PR. Since there are database changes you will need to run those as well. 

For MySQL users run this on your database:
```
INSERT IGNORE INTO `#__mail_templates` (`template_id`, `extension`, `language`, `subject`, `body`, `htmlbody`, `attachments`, `params`) VALUES
('plg_workflow_notification.mail', 'plg_workflow_notification', '', 'PLG_WORKFLOW_NOTIFICATION_EMAIL_ON_TRANSITION_SUBJECT', 'PLG_WORKFLOW_NOTIFICATION_EMAIL_ON_TRANSITION_MSG', '', '', '{"tags":["siteurl","title","performer","transitionName","toStage","extraText"]}');
```

For Postgres users run this on your database:
```
INSERT INTO "#__mail_templates" ("template_id", "extension", "language", "subject", "body", "htmlbody", "attachments", "params") VALUES
('plg_workflow_notification.mail', 'plg_workflow_notification', '', 'PLG_WORKFLOW_NOTIFICATION_EMAIL_ON_TRANSITION_SUBJECT', 'PLG_WORKFLOW_NOTIFICATION_EMAIL_ON_TRANSITION_MSG', '', '', '{"tags":["siteurl","title","performer","transitionName","toStage","extraText"]}')
ON CONFLICT DO NOTHING;
```

1. Enable Workflows by going to Content -> Articles -> Options -> Integration -> Enable Workflow and set it to Yes
2. Save the changes
3. Go to Content -> Workflows
4. Click on the number 7 (or another number) under Transitions
5. Click on the Unpublish transition
6. Click on Notification
7. Notice there is now the option to select a notification type
8. The `Private Messages` is the old and default way of sending notifications. Clicking on the dropdown now also let's you select `Email` as an option. Multiple options are possible as well
9. Select a usergroup and/or a user where you want to send the message to
10. Save the transition
11. Go got Content -> Articles
12. Edit an article
13. Set the transition to `Unpublished`
14. Save the article
15. Check if the message has arrived

#### Mail templates
The sending of the emails uses the mail templates in Joomla. The mail template can be customized via these steps:
1. Go to System -> Mail Templates
2. Filter on `workflow`
3. You will see 1 template. 
![image](https://user-images.githubusercontent.com/359377/169713800-c0c6ee73-6d92-4032-8780-a2000eab814a.png)
4. Click on the template to edit it
5. Modify the template to your needs

### Feedback requested
The following questions I have:
1. The name of the person doing the transition has a placeholder named `performer` Any suggestions for a better name?
2. Are there more placeholders wanted for the mail template?


@brianteeman Could you please check the language files?

#### Final words
Thanks to @bembelimen and @softforge for the idea to add this. Thanks to @brianteeman to point out to use mail templates.

I know this isn't perfect code so looking for feedback to get this into Joomla to make transitions more useful as they are quite neat.

### Actual result BEFORE applying this Pull Request
Messages are only sent to users who can receive private messages


### Expected result AFTER applying this Pull Request
Messages can now also be send to users with an account and email address.


### Documentation Changes Required
Yes, need to document the addition of the email option.
